File: android/app/src/main/java/com/noop/analytics/AdaptiveDayGuidance.kt
package com.noop.analytics

import java.time.Instant
import java.time.ZoneOffset
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Evidence-gated daily guidance. This is the Kotlin twin of AdaptiveDayGuidance.swift.
 *
 * It ranks observations but never infers why a routine changed. Missing or stale evidence fails closed.
 */
object AdaptiveDayGuidance {
    const val MINIMUM_ROUTINE_NIGHTS = 5
    const val STRONG_ROUTINE_NIGHTS = 7
    const val ROUTINE_LOOKBACK_DAYS = 21
    const val LATEST_SLEEP_MAXIMUM_AGE_SECONDS = 30 * 60 * 60
    const val SHORT_SLEEP_THRESHOLD_MINUTES = 60
    const val LATE_ROUTINE_THRESHOLD_MINUTES = 90
    const val ROUTINE_DURATION_DROP_MINUTES = 45
    const val MAXIMUM_FRAGMENT_GAP_SECONDS = 90L * 60L
    const val TRAVEL_THRESHOLD_SECONDS = 2 * 60 * 60
    const val TRAVEL_MAXIMUM_AGE_SECONDS = 36 * 60 * 60

    enum class Kind { TRAVEL_ADJUSTMENT, ROUTINE_RECOVERY, SLEEP_RECOVERY }
    enum class Confidence { BUILDING, STRONG }

    data class SleepDay(val day: String, val totalSleepMinutes: Double?)
    data class SleepWindow(val startSec: Long, val endSec: Long)
    data class TimeZoneChange(
        val previousOffsetSec: Int,
        val currentOffsetSec: Int,
        val observedAtSec: Long,
    )

    data class Input(
        val today: String,
        val nowSec: Long,
        val currentTimeZoneOffsetSec: Int,
        val sleepTargetMinutes: Int,
        val sleepTargetIsExplicit: Boolean,
        val sleepDays: List<SleepDay>,
        val sleepWindows: List<SleepWindow>,
        val timeZoneChange: TimeZoneChange?,
        val routineHistoryStartSec: Long? = null,
    )

    data class Recommendation(
        val kind: Kind,
        val observedAtSec: Long,
        val maximumAgeSeconds: Int,
        val confidence: Confidence,
        val fingerprint: String,
        val evidence: List<String>,
    )

    fun recommendation(input: Input): Recommendation? {
        travelRecommendation(input)?.let { return it }
        val windows = eligibleWindows(input)
        routineRecommendation(input, windows)?.let { return it }
        return sleepRecommendation(input, windows)
    }

    private fun travelRecommendation(input: Input): Recommendation? {
        val change = input.timeZoneChange ?: return null
        val age = input.nowSec - change.observedAtSec
        val delta = normalizedTravelDeltaSeconds(
            previousOffsetSec = change.previousOffsetSec,
            currentOffsetSec = change.currentOffsetSec,
        )
        if (
            age !in (-5 * 60L)..TRAVEL_MAXIMUM_AGE_SECONDS.toLong() ||
            abs(change.previousOffsetSec) > 14 * 60 * 60 ||
            abs(change.currentOffsetSec) > 14 * 60 * 60 ||
            abs(delta) < TRAVEL_THRESHOLD_SECONDS
        ) return null
        return Recommendation(
            kind = Kind.TRAVEL_ADJUSTMENT,
            observedAtSec = change.observedAtSec,
            maximumAgeSeconds = TRAVEL_MAXIMUM_AGE_SECONDS,
            confidence = Confidence.STRONG,
            fingerprint =
                "travel:${change.previousOffsetSec}:${change.currentOffsetSec}:${change.observedAtSec / 3600L}",
            evidence = listOf(if (delta > 0) "timezone-east" else "timezone-west", "offset-change"),
        )
    }

    /**
     * Shortest wall-clock shift between two UTC offsets. A raw 22-hour Date Line change is the same
     * circadian shift as two hours; a raw 24-hour change leaves the local clock unchanged.
     */
    fun normalizedTravelDeltaSeconds(previousOffsetSec: Int, currentOffsetSec: Int): Int {
        val day = 24 * 60 * 60
        val halfDay = day / 2
        val raw = currentOffsetSec - previousOffsetSec
        return ((raw + halfDay) % day + day) % day - halfDay
    }

    private data class SleepObservation(
        val primaryStartSec: Long,
        val endSec: Long,
        val totalDurationSeconds: Long,
    )

    private fun eligibleWindows(input: Input): List<SleepObservation> {
        val oldest = input.nowSec - ROUTINE_LOOKBACK_DAYS * 24L * 60L * 60L
        val seen = mutableSetOf<String>()
        val candidates = input.sleepWindows
            .filter { window ->
                val duration = window.endSec - window.startSec
                window.startSec >= oldest &&
                    (input.routineHistoryStartSec?.let { window.startSec >= it } != false) &&
                    window.endSec <= input.nowSec &&
                    duration in 1L..(14 * 60 * 60L) &&
                    seen.add("${window.startSec}:${window.endSec}")
            }
            .sortedWith(compareBy<SleepWindow> { it.startSec }.thenBy { it.endSec })

        // Merge short, nearby fragments before the three-hour night threshold. Distant sleep remains a
        // separate cluster, and only one cluster can represent a noon-to-noon sleep night.
        return candidates
            .groupBy { sleepNightKey(it.startSec, input.currentTimeZoneOffsetSec) }
            .values
            .mapNotNull { windows ->
                fragmentClusters(windows)
                    .mapNotNull { cluster ->
                        val startSec = cluster.minOfOrNull { it.startSec }
                            ?: return@mapNotNull null
                        if (!isOvernightOnset(startSec, input.currentTimeZoneOffsetSec)) {
                            return@mapNotNull null
                        }
                        val totalDuration = mergedDurationSeconds(cluster)
                        if (totalDuration !in (3L * 60L * 60L)..(14L * 60L * 60L)) {
                            return@mapNotNull null
                        }
                        SleepObservation(
                            primaryStartSec = startSec,
                            endSec = cluster.maxOf { it.endSec },
                            totalDurationSeconds = totalDuration,
                        )
                    }
                    .maxWithOrNull(
                        compareBy<SleepObservation> { it.totalDurationSeconds }.thenBy { it.endSec },
                    )
            }
            .sortedBy { it.endSec }
    }

    private fun routineRecommendation(
        input: Input,
        windows: List<SleepObservation>,
    ): Recommendation? {
        val latest = windows.lastOrNull() ?: return null
        if (input.nowSec - latest.endSec > LATEST_SLEEP_MAXIMUM_AGE_SECONDS) return null
        val history = windows.dropLast(1).takeLast(14)
        if (history.size < MINIMUM_ROUTINE_NIGHTS) return null

        val baselineOnset = median(
            history.map {
                bedtimeCoordinate(it.primaryStartSec, input.currentTimeZoneOffsetSec)
            },
        )
        val latestOnset = bedtimeCoordinate(
            latest.primaryStartSec,
            input.currentTimeZoneOffsetSec,
        )
        val delay = latestOnset - baselineOnset
        val baselineDuration = median(history.map { it.totalDurationSeconds / 60.0 })
        val latestDuration = latest.totalDurationSeconds / 60.0
        val shortened = latestDuration <= baselineDuration - ROUTINE_DURATION_DROP_MINUTES
        if (delay < LATE_ROUTINE_THRESHOLD_MINUTES || !shortened) return null

        return Recommendation(
            kind = Kind.ROUTINE_RECOVERY,
            observedAtSec = latest.endSec,
            maximumAgeSeconds = 18 * 60 * 60,
            confidence = if (history.size >= STRONG_ROUTINE_NIGHTS) {
                Confidence.STRONG
            } else {
                Confidence.BUILDING
            },
            fingerprint =
                "routine:${latest.primaryStartSec / 900L}:${latest.endSec / 900L}:" +
                    "${latest.totalDurationSeconds / 900L}",
            evidence = listOf("personal-sleep-timing", "later-onset", "shorter-sleep"),
        )
    }

    private fun sleepRecommendation(
        input: Input,
        windows: List<SleepObservation>,
    ): Recommendation? {
        if (!input.sleepTargetIsExplicit) return null
        val target = input.sleepTargetMinutes.coerceIn(5 * 60, 11 * 60).toDouble()
        val freshWindows = windows.asReversed().filter { window ->
            input.nowSec - window.endSec in
                (-5 * 60L)..LATEST_SLEEP_MAXIMUM_AGE_SECONDS.toLong() &&
                localDayKey(window.endSec, input.currentTimeZoneOffsetSec) == input.today
        }
        val current = input.sleepDays.lastOrNull { it.day == input.today }
        val minutes = current?.totalSleepMinutes
        val matchedWindow = if (
            minutes != null &&
            minutes.isFinite() &&
            minutes in 120.0..900.0 &&
            target - minutes >= SHORT_SLEEP_THRESHOLD_MINUTES
        ) {
            freshWindows.firstOrNull { window ->
                DailyActionPlanner.matchedSleepObservationEndSec(
                    aggregateMinutes = minutes,
                    sessionDurationMinutes = window.totalDurationSeconds / 60.0,
                    sessionEndSec = window.endSec,
                    nowSec = input.nowSec,
                    maximumAgeSeconds = LATEST_SLEEP_MAXIMUM_AGE_SECONDS,
                ) != null
            }
        } else {
            null
        }
        if (
            minutes != null &&
            matchedWindow != null
        ) {
            return Recommendation(
                kind = Kind.SLEEP_RECOVERY,
                observedAtSec = matchedWindow.endSec,
                maximumAgeSeconds = 18 * 60 * 60,
                confidence = Confidence.STRONG,
                fingerprint = "sleep:${input.today}:${minutes.roundToInt()}",
                evidence = listOf("current-sleep", "below-explicit-target"),
            )
        }

        val latest = freshWindows.firstOrNull() ?: return null
        val duration = latest.totalDurationSeconds / 60.0
        if (target - duration < SHORT_SLEEP_THRESHOLD_MINUTES) return null
        return Recommendation(
            kind = Kind.SLEEP_RECOVERY,
            observedAtSec = latest.endSec,
            maximumAgeSeconds = 18 * 60 * 60,
            confidence = Confidence.BUILDING,
            fingerprint =
                "sleep-window:${latest.primaryStartSec / 900L}:${latest.endSec / 900L}:" +
                    "${latest.totalDurationSeconds / 900L}",
            evidence = listOf("recent-sleep-window", "below-explicit-target"),
        )
    }

    private fun fragmentClusters(windows: List<SleepWindow>): List<List<SleepWindow>> {
        val sorted = windows.sortedWith(compareBy<SleepWindow> { it.startSec }.thenBy { it.endSec })
        val clusters = mutableListOf<MutableList<SleepWindow>>()
        for (window in sorted) {
            val last = clusters.lastOrNull()
            val clusterEnd = last?.maxOfOrNull { it.endSec }
            if (last == null || clusterEnd == null ||
                window.startSec - clusterEnd > MAXIMUM_FRAGMENT_GAP_SECONDS
            ) {
                clusters += mutableListOf(window)
            } else {
                last += window
            }
        }
        return clusters
    }

    private fun mergedDurationSeconds(windows: List<SleepWindow>): Long {
        val sorted = windows.sortedWith(compareBy<SleepWindow> { it.startSec }.thenBy { it.endSec })
        val first = sorted.firstOrNull() ?: return 0L
        var currentStart = first.startSec
        var currentEnd = first.endSec
        var total = 0L
        for (window in sorted.drop(1)) {
            if (window.startSec <= currentEnd) {
                currentEnd = maxOf(currentEnd, window.endSec)
            } else {
                total += currentEnd - currentStart
                currentStart = window.startSec
                currentEnd = window.endSec
            }
        }
        return total + currentEnd - currentStart
    }

    private fun localDayKey(epochSec: Long, offsetSec: Int): String =
        Instant.ofEpochSecond(epochSec + offsetSec)
            .atOffset(ZoneOffset.UTC)
            .toLocalDate()
            .toString()

    private fun isOvernightOnset(epochSec: Long, offsetSec: Int): Boolean {
        val minute = localMinute(epochSec, offsetSec)
        return minute >= 18 * 60 || minute < 6 * 60
    }

    private fun sleepNightKey(epochSec: Long, offsetSec: Int): Long =
        Math.floorDiv(epochSec + offsetSec - 12L * 60L * 60L, 24L * 60L * 60L)

    private fun bedtimeCoordinate(epochSec: Long, offsetSec: Int): Double {
        val minute = localMinute(epochSec, offsetSec)
        return (if (minute < 12 * 60) minute + 24 * 60 else minute).toDouble()
    }

    private fun localMinute(epochSec: Long, offsetSec: Int): Int {
        val day = 24 * 60 * 60L
        val local = ((epochSec + offsetSec) % day + day) % day
        return (local / 60L).toInt()
    }

    private fun median(values: List<Double>): Double {
        val sorted = values.sorted()
        if (sorted.isEmpty()) return 0.0
        val mid = sorted.size / 2
        return if (sorted.size % 2 == 0) {
            (sorted[mid - 1] + sorted[mid]) / 2.0
        } else {
            sorted[mid]
        }
    }
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/AdaptiveDayGuidance.swift
import Foundation

/// Evidence-gated daily guidance shared by the Apple and Android apps.
///
/// This engine ranks observations; it does not schedule notifications or infer causes. In particular,
/// a late night is described as a routine shift, never as alcohol, a party, illness, or poor judgment.
/// Missing or stale data fails closed.
public enum AdaptiveDayGuidance {
    public static let minimumRoutineNights = 5
    public static let strongRoutineNights = 7
    public static let routineLookbackDays = 21
    public static let latestSleepMaximumAgeSeconds = 30 * 60 * 60
    public static let shortSleepThresholdMinutes = 60
    public static let lateRoutineThresholdMinutes = 90
    public static let routineDurationDropMinutes = 45
    public static let maximumFragmentGapSeconds = 90 * 60
    public static let travelThresholdSeconds = 2 * 60 * 60
    public static let travelMaximumAgeSeconds = 36 * 60 * 60

    public enum Kind: String, Equatable, Sendable, Codable {
        case travelAdjustment
        case routineRecovery
        case sleepRecovery
    }

    public enum Confidence: String, Equatable, Sendable, Codable {
        case building
        case strong
    }

    public struct SleepDay: Equatable, Sendable {
        public let day: String
        public let totalSleepMinutes: Double?

        public init(day: String, totalSleepMinutes: Double?) {
            self.day = day
            self.totalSleepMinutes = totalSleepMinutes
        }
    }

    public struct SleepWindow: Equatable, Sendable {
        public let startSec: Int
        public let endSec: Int

        public init(startSec: Int, endSec: Int) {
            self.startSec = startSec
            self.endSec = endSec
        }
    }

    /// A persisted timezone transition. Keeping the observation after the offset snapshot advances lets
    /// delivery retry after quiet hours or a temporarily unavailable notification permission.
    public struct TimeZoneChange: Equatable, Sendable {
        public let previousOffsetSec: Int
        public let currentOffsetSec: Int
        public let observedAtSec: Int

        public init(previousOffsetSec: Int, currentOffsetSec: Int, observedAtSec: Int) {
            self.previousOffsetSec = previousOffsetSec
            self.currentOffsetSec = currentOffsetSec
            self.observedAtSec = observedAtSec
        }
    }

    public struct Input: Equatable, Sendable {
        public let today: String
        public let nowSec: Int
        public let currentTimeZoneOffsetSec: Int
        public let sleepTargetMinutes: Int
        public let sleepTargetIsExplicit: Bool
        public let sleepDays: [SleepDay]
        public let sleepWindows: [SleepWindow]
        public let timeZoneChange: TimeZoneChange?
        public let routineHistoryStartSec: Int?

        public init(
            today: String,
            nowSec: Int,
            currentTimeZoneOffsetSec: Int,
            sleepTargetMinutes: Int,
            sleepTargetIsExplicit: Bool,
            sleepDays: [SleepDay],
            sleepWindows: [SleepWindow],
            timeZoneChange: TimeZoneChange?,
            routineHistoryStartSec: Int? = nil
        ) {
            self.today = today
            self.nowSec = nowSec
            self.currentTimeZoneOffsetSec = currentTimeZoneOffsetSec
            self.sleepTargetMinutes = sleepTargetMinutes
            self.sleepTargetIsExplicit = sleepTargetIsExplicit
            self.sleepDays = sleepDays
            self.sleepWindows = sleepWindows
            self.timeZoneChange = timeZoneChange
            self.routineHistoryStartSec = routineHistoryStartSec
        }
    }

    public struct Recommendation: Equatable, Sendable {
        public let kind: Kind
        public let observedAtSec: Int
        public let maximumAgeSeconds: Int
        public let confidence: Confidence
        public let fingerprint: String
        /// Stable, value-free reason tokens for diagnostics and tests.
        public let evidence: [String]

        public init(
            kind: Kind,
            observedAtSec: Int,
            maximumAgeSeconds: Int,
            confidence: Confidence,
            fingerprint: String,
            evidence: [String]
        ) {
            self.kind = kind
            self.observedAtSec = observedAtSec
            self.maximumAgeSeconds = maximumAgeSeconds
            self.confidence = confidence
            self.fingerprint = fingerprint
            self.evidence = evidence
        }
    }

    /// Returns the single highest-priority current recommendation: travel, then a measured routine shift,
    /// then short-sleep guidance. Delivery-level cooldowns remain the app shell's responsibility.
    public static func recommendation(_ input: Input) -> Recommendation? {
        if let travel = travelRecommendation(input) { return travel }

        let windows = eligibleWindows(input)
        if let routine = routineRecommendation(input, windows: windows) { return routine }
        return sleepRecommendation(input, windows: windows)
    }

    private static func travelRecommendation(_ input: Input) -> Recommendation? {
        guard let change = input.timeZoneChange else { return nil }
        let age = input.nowSec - change.observedAtSec
        let delta = normalizedTravelDeltaSeconds(
            previousOffsetSec: change.previousOffsetSec,
            currentOffsetSec: change.currentOffsetSec
        )
        guard age >= -5 * 60,
              age <= travelMaximumAgeSeconds,
              abs(change.previousOffsetSec) <= 14 * 60 * 60,
              abs(change.currentOffsetSec) <= 14 * 60 * 60,
              abs(delta) >= travelThresholdSeconds else { return nil }
        return Recommendation(
            kind: .travelAdjustment,
            observedAtSec: change.observedAtSec,
            maximumAgeSeconds: travelMaximumAgeSeconds,
            confidence: .strong,
            fingerprint: "travel:\(change.previousOffsetSec):\(change.currentOffsetSec):\(change.observedAtSec / 3600)",
            evidence: [delta > 0 ? "timezone-east" : "timezone-west", "offset-change"]
        )
    }

    /// Return the shortest wall-clock shift between two UTC offsets. Raw offset subtraction reports a
    /// 22-hour change for a two-hour International Date Line crossing; circadian guidance should use
    /// the equivalent two-hour shift instead. Exactly 24 hours means the local clock did not move.
    public static func normalizedTravelDeltaSeconds(
        previousOffsetSec: Int,
        currentOffsetSec: Int
    ) -> Int {
        let day = 24 * 60 * 60
        let halfDay = day / 2
        let raw = currentOffsetSec - previousOffsetSec
        return ((raw + halfDay) % day + day) % day - halfDay
    }

    private struct SleepObservation {
        let primaryStartSec: Int
        let endSec: Int
        let totalDurationSeconds: Int
    }

    private static func eligibleWindows(_ input: Input) -> [SleepObservation] {
        let oldest = input.nowSec - routineLookbackDays * 24 * 60 * 60
        var seen = Set<String>()
        let candidates = input.sleepWindows
            .filter { window in
                let duration = window.endSec - window.startSec
                guard window.startSec >= oldest,
                      (input.routineHistoryStartSec.map {
                          window.startSec >= $0
                      } ?? true),
                      window.endSec <= input.nowSec,
                      duration > 0,
                      duration <= 14 * 60 * 60,
                      seen.insert("\(window.startSec):\(window.endSec)").inserted
                else { return false }
                return true
            }
            .sorted {
                $0.startSec == $1.startSec
                    ? $0.endSec < $1.endSec
                    : $0.startSec < $1.startSec
            }

        // Merge short, nearby fragments before applying the three-hour night threshold. Distant sleep
        // remains a separate cluster, and only one cluster can represent a noon-to-noon sleep night.
        var windowsByNight: [Int: [SleepWindow]] = [:]
        for window in candidates {
            let key = sleepNightKey(window.startSec, offsetSec: input.currentTimeZoneOffsetSec)
            windowsByNight[key, default: []].append(window)
        }
        return windowsByNight.values.compactMap { windows in
            fragmentClusters(windows).compactMap { cluster -> SleepObservation? in
                guard let startSec = cluster.map(\.startSec).min(),
                      let endSec = cluster.map(\.endSec).max(),
                      isOvernightOnset(
                          startSec,
                          offsetSec: input.currentTimeZoneOffsetSec
                      ) else { return nil }
                let totalDuration = mergedDurationSeconds(cluster)
                guard totalDuration >= 3 * 60 * 60,
                      totalDuration <= 14 * 60 * 60 else { return nil }
                return SleepObservation(
                    primaryStartSec: startSec,
                    endSec: endSec,
                    totalDurationSeconds: totalDuration
                )
            }
            .max {
                $0.totalDurationSeconds == $1.totalDurationSeconds
                    ? $0.endSec < $1.endSec
                    : $0.totalDurationSeconds < $1.totalDurationSeconds
            }
        }
        .sorted { $0.endSec < $1.endSec }
    }

    private static func routineRecommendation(
        _ input: Input,
        windows: [SleepObservation]
    ) -> Recommendation? {
        guard let latest = windows.last,
              input.nowSec - latest.endSec <= latestSleepMaximumAgeSeconds else { return nil }
        let history = Array(windows.dropLast().suffix(14))
        guard history.count >= minimumRoutineNights else { return nil }

        let baselineOnset = median(history.map {
            bedtimeCoordinate($0.primaryStartSec, offsetSec: input.currentTimeZoneOffsetSec)
        })
        let latestOnset = bedtimeCoordinate(
            latest.primaryStartSec,
            offsetSec: input.currentTimeZoneOffsetSec
        )
        let delay = latestOnset - baselineOnset
        let baselineDuration = median(history.map {
            Double($0.totalDurationSeconds) / 60.0
        })
        let latestDuration = Double(latest.totalDurationSeconds) / 60.0
        let shortened = latestDuration <= baselineDuration - Double(routineDurationDropMinutes)
        guard delay >= Double(lateRoutineThresholdMinutes), shortened else { return nil }

        return Recommendation(
            kind: .routineRecovery,
            observedAtSec: latest.endSec,
            maximumAgeSeconds: 18 * 60 * 60,
            confidence: history.count >= strongRoutineNights ? .strong : .building,
            fingerprint:
                "routine:\(latest.primaryStartSec / 900):\(latest.endSec / 900):" +
                "\(latest.totalDurationSeconds / 900)",
            evidence: ["personal-sleep-timing", "later-onset", "shorter-sleep"]
        )
    }

    private static func sleepRecommendation(
        _ input: Input,
        windows: [SleepObservation]
    ) -> Recommendation? {
        guard input.sleepTargetIsExplicit else { return nil }
        let target = Double(min(max(input.sleepTargetMinutes, 5 * 60), 11 * 60))
        let freshWindows = windows.reversed().filter { window in
            let age = input.nowSec - window.endSec
            return (-5 * 60...latestSleepMaximumAgeSeconds).contains(age)
                && localDayKey(
                    window.endSec,
                    offsetSec: input.currentTimeZoneOffsetSec
                ) == input.today
        }
        if let current = input.sleepDays.last(where: { $0.day == input.today }),
           let minutes = current.totalSleepMinutes,
           minutes.isFinite,
           (120...900).contains(minutes),
           target - minutes >= Double(shortSleepThresholdMinutes),
           let matchedWindow = freshWindows.first(where: { window in
               DailyActionPlanner.matchedSleepObservationEndSec(
                   aggregateMinutes: minutes,
                   sessionDurationMinutes: Double(window.totalDurationSeconds) / 60.0,
                   sessionEndSec: window.endSec,
                   nowSec: input.nowSec,
                   maximumAgeSeconds: latestSleepMaximumAgeSeconds
               ) != nil
           }) {
            return Recommendation(
                kind: .sleepRecovery,
                observedAtSec: matchedWindow.endSec,
                maximumAgeSeconds: 18 * 60 * 60,
                confidence: .strong,
                fingerprint: "sleep:\(input.today):\(Int(minutes.rounded()))",
                evidence: ["current-sleep", "below-explicit-target"]
            )
        }

        // A sleep block is a weaker fallback because time in bed is not the same as asleep time.
        guard let latest = freshWindows.first else { return nil }
        let duration = Double(latest.totalDurationSeconds) / 60.0
        guard target - duration >= Double(shortSleepThresholdMinutes) else { return nil }
        return Recommendation(
            kind: .sleepRecovery,
            observedAtSec: latest.endSec,
            maximumAgeSeconds: 18 * 60 * 60,
            confidence: .building,
            fingerprint:
                "sleep-window:\(latest.primaryStartSec / 900):\(latest.endSec / 900):" +
                "\(latest.totalDurationSeconds / 900)",
            evidence: ["recent-sleep-window", "below-explicit-target"]
        )
    }

    private static func fragmentClusters(_ windows: [SleepWindow]) -> [[SleepWindow]] {
        let sorted = windows.sorted {
            $0.startSec == $1.startSec
                ? $0.endSec < $1.endSec
                : $0.startSec < $1.startSec
        }
        var clusters: [[SleepWindow]] = []
        for window in sorted {
            guard let lastCluster = clusters.last,
                  let clusterEnd = lastCluster.map(\.endSec).max(),
                  window.startSec - clusterEnd <= maximumFragmentGapSeconds else {
                clusters.append([window])
                continue
            }
            clusters[clusters.count - 1].append(window)
        }
        return clusters
    }

    private static func mergedDurationSeconds(_ windows: [SleepWindow]) -> Int {
        let sorted = windows.sorted {
            $0.startSec == $1.startSec
                ? $0.endSec < $1.endSec
                : $0.startSec < $1.startSec
        }
        guard var current = sorted.first else { return 0 }
        var total = 0
        for window in sorted.dropFirst() {
            if window.startSec <= current.endSec {
                current = SleepWindow(
                    startSec: current.startSec,
                    endSec: max(current.endSec, window.endSec)
                )
            } else {
                total += current.endSec - current.startSec
                current = window
            }
        }
        return total + current.endSec - current.startSec
    }

    private static func localDayKey(_ epochSec: Int, offsetSec: Int) -> String {
        let date = Date(timeIntervalSince1970: TimeInterval(epochSec + offsetSec))
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    private static func isOvernightOnset(_ epochSec: Int, offsetSec: Int) -> Bool {
        let minute = localMinute(epochSec, offsetSec: offsetSec)
        return minute >= 18 * 60 || minute < 6 * 60
    }

    private static func sleepNightKey(_ epochSec: Int, offsetSec: Int) -> Int {
        Int(floor(Double(epochSec + offsetSec - 12 * 60 * 60) / Double(24 * 60 * 60)))
    }

    /// Maps evening through early-morning bedtimes onto one monotonic 18:00...30:00 axis.
    private static func bedtimeCoordinate(_ epochSec: Int, offsetSec: Int) -> Double {
        let minute = localMinute(epochSec, offsetSec: offsetSec)
        return Double(minute < 12 * 60 ? minute + 24 * 60 : minute)
    }

    private static func localMinute(_ epochSec: Int, offsetSec: Int) -> Int {
        let day = 24 * 60 * 60
        let local = ((epochSec + offsetSec) % day + day) % day
        return local / 60
    }

    private static func median(_ values: [Double]) -> Double {
        let sorted = values.sorted()
        guard !sorted.isEmpty else { return 0 }
        let mid = sorted.count / 2
        return sorted.count.isMultiple(of: 2)
            ? (sorted[mid - 1] + sorted[mid]) / 2
            : sorted[mid]
    }
}
